# Assignment 2. Data Science

This is the first assigment for the course. The idea is to get a deep understanding ofo Pandas library and get used to, in order to apply the same metodology for future work for _reading_, _manipulating_ and _cleaning_ datasets for **Machine Learning**.


### Part1: All Time Olympic Games Medals
Load the data set csv from "olympics.csv" using Pandas library.

First we need to import the library and load the file with the parameters we need in order to skip or set the way we the file is structured. Following the file looks like following

    0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15
    ,№ Summer,01 !,02 !,03 !,Total,№ Winter,01 !,02 !,03 !,Total,№ Games,01 !,02 !,03 !,Combined total
    Afghanistan (AFG),13,0,0,2,2,0,0,0,0,0,13,0,0,2,2
    Algeria (ALG),12,5,2,8,15,3,0,0,0,0,15,5,2,8,15
    Argentina (ARG),23,18,24,28,70,18,0,0,0,0,41,18,24,28,70
    Armenia (ARM),5,1,2,9,12,6,0,0,0,0,11,1,2,9,12

Seems we need to skip the first row, the other elements seems being accordingly to csv standard format.


In [1]:
import pandas as pd

In [221]:
df = pd.read_csv('olympics.csv', index_col=0, skiprows=1)

For the indexes we will be using the first column that it's the country.

Print the head of the read csv to see the results.

In [3]:
df.head()

№ Summer  01 !  02 !  03 !  Total  № Winter  01 !.1  \
Afghanistan (AFG)              13     0     0     2      2         0       0   
Algeria (ALG)                  12     5     2     8     15         3       0   
Argentina (ARG)                23    18    24    28     70        18       0   
Armenia (ARM)                   5     1     2     9     12         6       0   
Australasia (ANZ) [ANZ]         2     3     4     5     12         0       0   

                         02 !.1  03 !.1  Total.1  № Games  01 !.2  02 !.2  \
Afghanistan (AFG)             0       0        0       13       0       0   
Algeria (ALG)                 0       0        0       15       5       2   
Argentina (ARG)               0       0        0       41      18      24   
Armenia (ARM)                 0       0        0       11       1       2   
Australasia (ANZ) [ANZ]       0       0        0        2       3       4   

                         03 !.2  Combined total  
Afghanistan (AFG)             2               2  
Algeria (ALG)                 8              15  
Argentina (ARG)              28              70  
Armenia (ARM)                 9              12  
Australasia (ANZ) [ANZ]       5              12

Let's do what the assigment initially do with the dataframe

In [222]:
for col in df.columns:
    if col[:2]=='01':
        df.rename(columns={col:'Gold'+col[4:]}, inplace=True)
    if col[:2]=='02':
        df.rename(columns={col:'Silver'+col[4:]}, inplace=True)
    if col[:2]=='03':
        df.rename(columns={col:'Bronze'+col[4:]}, inplace=True)
    if col[:1]=='№':
        df.rename(columns={col:'#'+col[1:]}, inplace=True)

names_ids = df.index.str.split('\s\(') # split the index by '('

df.index = names_ids.str[0] # the [0] element is the country name (new index) 
df['ID'] = names_ids.str[1].str[:3] # the [1] element is the abbreviation or ID (take first 3 characters from that)

df = df.drop('Totals')
df.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Afghanistan        13     0       0       2      2         0       0   
Algeria            12     5       2       8     15         3       0   
Argentina          23    18      24      28     70        18       0   
Armenia             5     1       2       9     12         6       0   
Australasia         2     3       4       5     12         0       0   

             Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  \
Afghanistan         0         0        0       13       0         0         2   
Algeria             0         0        0       15       5         2         8   
Argentina           0         0        0       41      18        24        28   
Armenia             0         0        0       11       1         2         9   
Australasia         0         0        0        2       3         4         5   

             Combined total   ID  
Afghanistan               2  AFG  
Algeria                  15  ALG  
Argentina                70  ARG  
Armenia                  12  ARM  
Australasia              12  ANZ

It's seems they are redefining some of the columns names (**Gold**, **Silver**, **Bronze**, etc..) and the countries names, so the are more suitable for managing by using indexing in python. This is to map each medal type depending on the session, so they will have unique columns name for indexing.

For the index, as we can see in the table above, we have the country names. As for the new **"ID"** column, the code of the country, within the brackets, it's being used. Also the column called total is deleted.


### Question 0

Let's we start with the initial (example) question. I have to create a function per question, that will return the expected value.


In [5]:
# You should write your whole answer within the function provided. The autograder will call
# this function and compare the return value against the correct solution value
def answer_zero():
    # This function returns the row for Afghanistan, which is a Series object. The assignment
    # question description will tell you the general format the autograder is expecting
    return df.iloc[0]

# You can examine what your function returns by calling it in the cell. If you have questions
# about the assignment formats, check out the discussion forums for any FAQs
answer_zero() 

# Summer           13
Gold                0
Silver              0
Bronze              2
Total               2
# Winter            0
Gold.1              0
Silver.1            0
Bronze.1            0
Total.1             0
# Games            13
Gold.2              0
Silver.2            0
Bronze.2            2
Combined total      2
ID                AFG
Name: Afghanistan, dtype: object

In [6]:
# You should write your whole answer within the function provided. The autograder will call
# this function and compare the return value against the correct solution value
def answer_zero():
    # This function returns the row for Afghanistan, which is a Series object. The assignment
    # question description will tell you the general format the autograder is expecting
    return df.loc["Afghanistan"]

# You can examine what your function returns by calling it in the cell. If you have questions
# about the assignment formats, check out the discussion forums for any FAQs
answer_zero() 

# Summer           13
Gold                0
Silver              0
Bronze              2
Total               2
# Winter            0
Gold.1              0
Silver.1            0
Bronze.1            0
Total.1             0
# Games            13
Gold.2              0
Silver.2            0
Bronze.2            2
Combined total      2
ID                AFG
Name: Afghanistan, dtype: object

### Question 1

Which country has won the most gold medals in summer games?

This function should return a single string value.



In [7]:
def answer_one():
    return "YOUR ANSWER HERE"

In order to return the country with most gold medals in summer games, we have to get the argmax of the values in the *"Gold"* column, from the _summer games session_. 

First let's we take only the column we need by using indexing. In order to do that we are going to use dataframe indexing so  it's going to return a new view the the index and the column we need. Then we only have to apply a Numpy function that will return the max value. Then we hav to return the country name, that's it's going to be the index[0].

In [8]:
gms_df = df["Gold"]
gms_df.head()

Afghanistan     0
Algeria         5
Argentina      18
Armenia         1
Australasia     3
Name: Gold, dtype: int64

In [9]:
gms_argmax = df["Gold"].argmax()
print(gms_argmax)

United States


In [10]:
print(df["Gold"]["United States"])

976


We know that it's **"United states"** with **976** medals. Also we can sort the values within the dataframe by using that column values.

### Answer Question 1

In [11]:
def answer_one():
    return df["Gold"].argmax()
answer_one()

'United States'

### Question 2

Which country had the biggest difference between their summer and winter gold medal counts?

This function should return a single string value.


In [12]:
def answer_two():
    return "YOUR ANSWER HERE"

In order to get the biggest difference between summer and winter gold medals counts, we have to create additional column with this value in a view, and then use the same method we used in the question Nº 1.

In [13]:
gold_sw = df[["Gold","Gold.1"]]
gold_sw.head()

Gold  Gold.1
Afghanistan     0       0
Algeria         5       0
Argentina      18       0
Armenia         1       0
Australasia     3       0

We can create additional columns to this view by using following sentence:

In [14]:
gold_sw["diff"] = gold_sw["Gold"] - gold_sw["Gold.1"]
gold_sw.head()

C:\Users\javier.santos\AppData\Local\Continuum\Anaconda3\lib\site-packages\ipykernel\__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


Gold  Gold.1  diff
Afghanistan     0       0     0
Algeria         5       0     5
Argentina      18       0    18
Armenia         1       0     1
Australasia     3       0     3

In [15]:
gold_sw = gold_sw.loc[:,"Gold"] - gold_sw.loc[:,"Gold.1"]
gold_sw.head()

Afghanistan     0
Algeria         5
Argentina      18
Armenia         1
Australasia     3
dtype: int64

In [16]:
print(gold_sw.argmax())

United States


In [17]:
print(gold_sw.loc["United States"])

# Verify whether the values are correct
print(df.loc["United States"])
print(df.loc["United States","Gold"])
print(df.loc["United States","Gold.1"])


880
# Summer            26
Gold               976
Silver             757
Bronze             666
Total             2399
# Winter            22
Gold.1              96
Silver.1           102
Bronze.1            84
Total.1            282
# Games             48
Gold.2            1072
Silver.2           859
Bronze.2           750
Combined total    2681
ID                 USA
Name: United States, dtype: object
976
96


In [18]:
def answer_two():
    return (df.loc[:,"Gold"] - df.loc[:,"Gold.1"]).argmax()
answer_two()

'United States'

### Question 3
Which country has the biggest difference between their summer gold medal counts and winter gold medal counts relative to their total gold medal count? 

$$\frac{Summer~Gold - Winter~Gold}{Total~Gold}$$

Only include countries that have won at least 1 gold in both summer and winter.

*This function should return a single string value.*

First we can how can we get the total of a column.

In [19]:
df["Gold"].sum()

4809

In [20]:
print("*****Substract golds medals: Summer - Winter******")
num = df.loc[:,"Gold"] - df.loc[:,"Gold.1"]
print(num.head())
print("****Sorten values******")
df_order = df.sort_values(["Gold"],axis=0, ascending=False)
print(df_order.head())
print(df_order.tail())
print("****Total Gold******")
den = df.loc[:,"Gold"] + df.loc[:,"Gold.1"]
print(den.loc["United States"])
print(den.loc["Great Britain"])
print("***TOTAL SUM")
print(den.sum())

*****Substract golds medals: Summer - Winter******
Afghanistan     0
Algeria         5
Argentina      18
Armenia         1
Australasia     3
dtype: int64
****Sorten values******
               # Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
United States        26   976     757     666   2399        22      96   
Soviet Union          9   395     319     296   1010         9      78   
Great Britain        27   236     272     272    780        22      10   
France               27   202     223     246    671        22      31   
China                 9   201     146     126    473        10      12   

               Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  \
United States       102        84      282       48    1072       859   
Soviet Union         57        59      194       18     473       376   
Great Britain         4        12       26       49     246       276   
France               31        47      109       49     233       254   
China       

In [21]:
print("****Total Gold (No compute)******")
print(df["Gold.2"].sum())
print("****Total Gold (No compute)******")
op = num/df["Gold.2"].sum()
print(op)
print("*** SOLUTION ***")
print(op.argmax())

****Total Gold (No compute)******
5768
****Total Gold (No compute)******
Afghanistan                         0.000000
Algeria                             0.000867
Argentina                           0.003121
Armenia                             0.000173
Australasia                         0.000520
Australia                           0.023232
Austria                            -0.007108
Azerbaijan                          0.001040
Bahamas                             0.000867
Bahrain                             0.000000
Barbados                            0.000000
Belarus                             0.001040
Belgium                             0.006241
Bermuda                             0.000000
Bohemia                             0.000000
Botswana                            0.000000
Brazil                              0.003988
British West Indies                 0.000000
Bulgaria                            0.008669
Burundi                             0.000173
Cameroon                   

### Answer Question 3

In [219]:
def answer_three():
    return ((df.loc[:,"Gold"] - df.loc[:,"Gold.1"])/df["Gold.2"].sum()).argmax()
answer_three()

'United States'

In [225]:
df.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Afghanistan        13     0       0       2      2         0       0   
Algeria            12     5       2       8     15         3       0   
Argentina          23    18      24      28     70        18       0   
Armenia             5     1       2       9     12         6       0   
Australasia         2     3       4       5     12         0       0   

             Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  \
Afghanistan         0         0        0       13       0         0         2   
Algeria             0         0        0       15       5         2         8   
Argentina           0         0        0       41      18        24        28   
Armenia             0         0        0       11       1         2         9   
Australasia         0         0        0        2       3         4         5   

             Combined total   ID  
Afghanistan               2  AFG  
Algeria                  15  ALG  
Argentina                70  ARG  
Armenia                  12  ARM  
Australasia              12  ANZ

In [230]:
filter = df[(df["Gold"]>0) & (df["Gold.1"]>0)]
filter.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  \
Australia        25   139     152     177    468        18       5         3   
Austria          26    18      33      35     86        22      59        78   
Belarus           5    12      24      39     75         6       6         4   
Belgium          25    37      52      53    142        20       1         1   
Bulgaria         19    51      85      78    214        19       1         2   

           Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  \
Australia         4       12       43     144       155       181   
Austria          81      218       48      77       111       116   
Belarus           5       15       11      18        28        44   
Belgium           3        5       45      38        53        56   
Bulgaria          3        6       38      52        87        81   

           Combined total   ID  
Australia             480  AUS  
Austria               304  AUT  
Belarus                90  BLR  
Belgium               147  BEL  
Bulgaria              220  BUL

Seems I left the sum function at the end. Since I am using the total value, no sum need to be computed

In [229]:
def answer_three():
    filter = df[(df["Gold"]>0) & (df["Gold.1"]>0)]
    return ((filter.loc[:,"Gold"] - filter.loc[:,"Gold.1"])/filter["Gold.2"]).argmax()
answer_three()

'Bulgaria'

In [218]:
new_loc = (df.loc[:,"Gold"] - df.loc[:,"Gold.1"])/df["Gold.2"]
new_loc

Afghanistan                              NaN
Algeria                             1.000000
Argentina                           1.000000
Armenia                             1.000000
Australasia                         1.000000
Australia                           0.930556
Austria                            -0.532468
Azerbaijan                          1.000000
Bahamas                             1.000000
Bahrain                                  NaN
Barbados                                 NaN
Belarus                             0.333333
Belgium                             0.947368
Bermuda                                  NaN
Bohemia                                  NaN
Botswana                                 NaN
Brazil                              1.000000
British West Indies                      NaN
Bulgaria                            0.961538
Burundi                             1.000000
Cameroon                            1.000000
Canada                             -0.024793
Chile     

### Question 4
Write a function that creates a Series called "Points" which is a weighted value where each gold medal (`Gold.2`) counts for 3 points, silver medals (`Silver.2`) for 2 points, and bronze medals (`Bronze.2`) for 1 point. The function should return only the column (a Series object) which you created.

*This function should return a Series named `Points` of length 146*

In [23]:
print(df.count())
print(type(df))
print(type(df["Gold"]))
print(type(df[["Gold","Gold.1"]]))
print(type(df.loc[:,["Gold","Gold.1"]]))

# Summer          146
Gold              146
Silver            146
Bronze            146
Total             146
# Winter          146
Gold.1            146
Silver.1          146
Bronze.1          146
Total.1           146
# Games           146
Gold.2            146
Silver.2          146
Bronze.2          146
Combined total    146
ID                146
dtype: int64
<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>


Indexing a single column it will return a Series, however selecting multiple columns will return a dataframe

In [24]:
df.loc[:,["Gold.2","Silver.2","Bronze.2"]].head()

Gold.2  Silver.2  Bronze.2
Afghanistan       0         0         2
Algeria           5         2         8
Argentina        18        24        28
Armenia           1         2         9
Australasia       3         4         5

In [25]:
new_df = (df["Gold.2"] * 3) + (df["Silver.2"]) * 2 + (df["Bronze.2"] * 1)
new_df.head()

Afghanistan      2
Algeria         27
Argentina      130
Armenia         16
Australasia     22
dtype: int64

In [26]:
print(type(new_df))
(18*3)+(24*2)+(28)

<class 'pandas.core.series.Series'>


130

In [27]:
new_df.rename("Points")

Afghanistan                            2
Algeria                               27
Argentina                            130
Armenia                               16
Australasia                           22
Australia                            923
Austria                              569
Azerbaijan                            43
Bahamas                               24
Bahrain                                1
Barbados                               1
Belarus                              154
Belgium                              276
Bermuda                                1
Bohemia                                5
Botswana                               2
Brazil                               184
British West Indies                    2
Bulgaria                             411
Burundi                                3
Cameroon                              12
Canada                               846
Chile                                 24
China                               1120
Colombia        

In [28]:
my_serie = pd.Series(new_df,name="Points")

In [29]:
my_serie.head()

Afghanistan      2
Algeria         27
Argentina      130
Armenia         16
Australasia     22
Name: Points, dtype: int64

### Answer Question 5

In [30]:
def answer_four():
    return pd.Series((df["Gold.2"] * 3) + (df["Silver.2"]) * 2 + df["Bronze.2"],name="Points")

In [31]:
answer_four().head()

Afghanistan      2
Algeria         27
Argentina      130
Armenia         16
Australasia     22
Name: Points, dtype: int64

### Part 2

For the next set of questions, we will be using census data from the United States Census Bureau. Counties are political and geographic subdivisions of states in the United States. This dataset contains population data for counties and states in the US from 2010 to 2015. See this document for a description of the variable names.

The census dataset (census.csv) should be loaded as census_df. Answer questions using this as appropriate.


### Question 5

Which state has the most counties in it? (hint: consider the sumlevel key carefully! You'll need this for future questions too...)

This function should return a single string value.


In [32]:
census_df = pd.read_csv('census.csv')
census_df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
0      40       3         6      1       0  Alabama         Alabama   
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
0        4779736            4780127          4785161     ...        
1          54571              54571            54660     ...        
2         182265             182265           183193     ...        
3          27457              27457            27341     ...        
4          22915              22919            22861     ...        

   RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
0          0.002295         -0.193196          0.381066          0.582002   
1          7.242091         -2.915927         -3.012349          2.265971   
2         14.832960         17.647293         21.845705         19.243287   
3         -4.728132         -2.500690         -7.056824         -3.904217   
4         -5.527043         -5.068871         -6.201001         -0.177537   

   RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
0         -0.467369     1.030015     0.826644     1.383282     1.724718   
1         -2.530799     7.606016    -2.626146    -2.722002     2.592270   
2         17.197872    15.844176    18.559627    22.727626    20.317142   
3        -10.543299    -4.874741    -2.758113    -7.167664    -3.978583   
4          0.177258    -5.088389    -4.363636    -5.403729     0.754533   

   RNETMIG2015  
0     0.712594  
1    -2.187333  
2    18.293499  
3   -10.543299  
4     1.107861  

[5 rows x 100 columns]

In [33]:
print(census_df["STNAME"].unique())

['Alabama' 'Alaska' 'Arizona' 'Arkansas' 'California' 'Colorado'
 'Connecticut' 'Delaware' 'District of Columbia' 'Florida' 'Georgia'
 'Hawaii' 'Idaho' 'Illinois' 'Indiana' 'Iowa' 'Kansas' 'Kentucky'
 'Louisiana' 'Maine' 'Maryland' 'Massachusetts' 'Michigan' 'Minnesota'
 'Mississippi' 'Missouri' 'Montana' 'Nebraska' 'Nevada' 'New Hampshire'
 'New Jersey' 'New Mexico' 'New York' 'North Carolina' 'North Dakota'
 'Ohio' 'Oklahoma' 'Oregon' 'Pennsylvania' 'Rhode Island' 'South Carolina'
 'South Dakota' 'Tennessee' 'Texas' 'Utah' 'Vermont' 'Virginia'
 'Washington' 'West Virginia' 'Wisconsin' 'Wyoming']


In [34]:
# Get the unique values for the state


filter = census_df["STNAME"] == "Alabama"
alabama_counties = census_df[filter]
print(alabama_counties.count())
print(alabama_counties.head())

SUMLEV                   68
REGION                   68
DIVISION                 68
STATE                    68
COUNTY                   68
STNAME                   68
CTYNAME                  68
CENSUS2010POP            68
ESTIMATESBASE2010        68
POPESTIMATE2010          68
POPESTIMATE2011          68
POPESTIMATE2012          68
POPESTIMATE2013          68
POPESTIMATE2014          68
POPESTIMATE2015          68
NPOPCHG_2010             68
NPOPCHG_2011             68
NPOPCHG_2012             68
NPOPCHG_2013             68
NPOPCHG_2014             68
NPOPCHG_2015             68
BIRTHS2010               68
BIRTHS2011               68
BIRTHS2012               68
BIRTHS2013               68
BIRTHS2014               68
BIRTHS2015               68
DEATHS2010               68
DEATHS2011               68
DEATHS2012               68
                         ..
RBIRTH2011               68
RBIRTH2012               68
RBIRTH2013               68
RBIRTH2014               68
RBIRTH2015          

In [35]:
census_df["STNAME"].value_counts()

Texas                   255
Georgia                 160
Virginia                134
Kentucky                121
Missouri                116
Kansas                  106
Illinois                103
North Carolina          101
Iowa                    100
Tennessee                96
Nebraska                 94
Indiana                  93
Ohio                     89
Minnesota                88
Michigan                 84
Mississippi              83
Oklahoma                 78
Arkansas                 76
Wisconsin                73
Pennsylvania             68
Alabama                  68
Florida                  68
South Dakota             67
Colorado                 65
Louisiana                65
New York                 63
California               59
Montana                  57
West Virginia            56
North Dakota             54
South Carolina           47
Idaho                    45
Washington               40
Oregon                   37
New Mexico               34
Alaska              

In [36]:
census_df["STNAME"].value_counts().argmax()

'Texas'

In [37]:
census_df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
0      40       3         6      1       0  Alabama         Alabama   
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
0        4779736            4780127          4785161     ...        
1          54571              54571            54660     ...        
2         182265             182265           183193     ...        
3          27457              27457            27341     ...        
4          22915              22919            22861     ...        

   RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
0          0.002295         -0.193196          0.381066          0.582002   
1          7.242091         -2.915927         -3.012349          2.265971   
2         14.832960         17.647293         21.845705         19.243287   
3         -4.728132         -2.500690         -7.056824         -3.904217   
4         -5.527043         -5.068871         -6.201001         -0.177537   

   RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
0         -0.467369     1.030015     0.826644     1.383282     1.724718   
1         -2.530799     7.606016    -2.626146    -2.722002     2.592270   
2         17.197872    15.844176    18.559627    22.727626    20.317142   
3        -10.543299    -4.874741    -2.758113    -7.167664    -3.978583   
4          0.177258    -5.088389    -4.363636    -5.403729     0.754533   

   RNETMIG2015  
0     0.712594  
1    -2.187333  
2    18.293499  
3   -10.543299  
4     1.107861  

[5 rows x 100 columns]

In [38]:
census_df_mod = census_df[census_df["SUMLEV"] != 40]
census_df_mod.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   
5      50       3         6      1       9  Alabama   Blount County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
1          54571              54571            54660     ...        
2         182265             182265           183193     ...        
3          27457              27457            27341     ...        
4          22915              22919            22861     ...        
5          57322              57322            57373     ...        

   RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
1          7.242091         -2.915927         -3.012349          2.265971   
2         14.832960         17.647293         21.845705         19.243287   
3         -4.728132         -2.500690         -7.056824         -3.904217   
4         -5.527043         -5.068871         -6.201001         -0.177537   
5          1.807375         -1.177622         -1.748766         -2.062535   

   RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
1         -2.530799     7.606016    -2.626146    -2.722002     2.592270   
2         17.197872    15.844176    18.559627    22.727626    20.317142   
3        -10.543299    -4.874741    -2.758113    -7.167664    -3.978583   
4          0.177258    -5.088389    -4.363636    -5.403729     0.754533   
5         -1.369970     1.859511    -0.848580    -1.402476    -1.577232   

   RNETMIG2015  
1    -2.187333  
2    18.293499  
3   -10.543299  
4     1.107861  
5    -0.884411  

[5 rows x 100 columns]

In [39]:
print(census_df.count())
print(census_df_mod.count())

SUMLEV                   3193
REGION                   3193
DIVISION                 3193
STATE                    3193
COUNTY                   3193
STNAME                   3193
CTYNAME                  3193
CENSUS2010POP            3193
ESTIMATESBASE2010        3193
POPESTIMATE2010          3193
POPESTIMATE2011          3193
POPESTIMATE2012          3193
POPESTIMATE2013          3193
POPESTIMATE2014          3193
POPESTIMATE2015          3193
NPOPCHG_2010             3193
NPOPCHG_2011             3193
NPOPCHG_2012             3193
NPOPCHG_2013             3193
NPOPCHG_2014             3193
NPOPCHG_2015             3193
BIRTHS2010               3193
BIRTHS2011               3193
BIRTHS2012               3193
BIRTHS2013               3193
BIRTHS2014               3193
BIRTHS2015               3193
DEATHS2010               3193
DEATHS2011               3193
DEATHS2012               3193
                         ... 
RBIRTH2011               3193
RBIRTH2012               3193
RBIRTH2013

In [40]:
result = census_df[census_df["SUMLEV"] != 40]["STNAME"].value_counts().argmax()
print(result)

Texas


### Answer Question 5

In [41]:
def answer_five():
    return census_df[census_df["SUMLEV"] != 40]["STNAME"].value_counts().argmax()
answer_five()

'Texas'

### Question 6

Only looking at the three most populous counties for each state, what are the three most populous states (in order of highest population to lowest population)? Use CENSUS2010POP.

This function should return a list of string values.

In [42]:
sort_list = census_df.sort_values(by="CENSUS2010POP", ascending=False)
sort_list.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY      STNAME     CTYNAME  \
190       40       4         9      6       0  California  California   
2566      40       3         7     48       0       Texas       Texas   
1860      40       1         2     36       0    New York    New York   
329       40       3         5     12       0     Florida     Florida   
608       40       2         3     17       0    Illinois    Illinois   

      CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
190        37253956           37254503         37334079     ...        
2566       25145561           25146105         25244363     ...        
1860       19378102           19378087         19402920     ...        
329        18801310           18804623         18849890     ...        
608        12830632           12831549         12841249     ...        

      RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
190          -1.148464         -1.163788         -1.339869         -0.862856   
2566          4.527491          5.484904          4.446476          6.072953   
1860         -4.354557         -5.945385         -5.561401         -8.130297   
329           5.556307          5.138184          4.847420          6.958576   
608          -5.243330         -5.805172         -5.308960         -7.509147   

      RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
190          -1.981572     2.761704     2.647127     2.728645     3.743342   
2566          6.248252     7.616364     8.660719     7.739851     9.780487   
1860         -7.990563     1.439239    -0.402860     0.407960    -1.334639   
329          10.080932    11.364173    10.676030    10.565882    13.478570   
608          -8.174675    -3.020410    -3.519051    -2.830379    -4.625216   

      RNETMIG2015  
190      2.656065  
2566     9.978697  
1860    -1.205321  
329     16.528676  
608     -5.247030  

[5 rows x 100 columns]

In [43]:
normalized_census =census_df[census_df["SUMLEV"] != 40]
three_countries = normalized_census.groupby('STNAME').head(3)
three_countries.head(20)

SUMLEV  REGION  DIVISION  STATE  COUNTY       STNAME  \
1        50       3         6      1       1      Alabama   
2        50       3         6      1       3      Alabama   
3        50       3         6      1       5      Alabama   
69       50       4         9      2      13       Alaska   
70       50       4         9      2      16       Alaska   
71       50       4         9      2      20       Alaska   
99       50       4         8      4       1      Arizona   
100      50       4         8      4       3      Arizona   
101      50       4         8      4       5      Arizona   
115      50       3         7      5       1     Arkansas   
116      50       3         7      5       3     Arkansas   
117      50       3         7      5       5     Arkansas   
191      50       4         9      6       1   California   
192      50       4         9      6       3   California   
193      50       4         9      6       5   California   
250      50       4         8      8       1     Colorado   
251      50       4         8      8       3     Colorado   
252      50       4         8      8       5     Colorado   
315      50       1         1      9       1  Connecticut   
316      50       1         1      9       3  Connecticut   

                        CTYNAME  CENSUS2010POP  ESTIMATESBASE2010  \
1                Autauga County          54571              54571   
2                Baldwin County         182265             182265   
3                Barbour County          27457              27457   
69       Aleutians East Borough           3141               3141   
70   Aleutians West Census Area           5561               5561   
71       Anchorage Municipality         291826             291826   
99                Apache County          71518              71518   
100              Cochise County         131346             131357   
101             Coconino County         134421             134437   
115             Arkansas County          19019              19018   
116               Ashley County          21853              21853   
117               Baxter County          41513              41513   
191              Alameda County        1510271            1510261   
192               Alpine County           1175               1175   
193               Amador County          38091              38091   
250                Adams County         441603             441687   
251              Alamosa County          15445              15445   
252             Arapahoe County         572003             572155   
315            Fairfield County         916829             916850   
316             Hartford County         894014             894029   

     POPESTIMATE2010     ...       RDOMESTICMIG2011  RDOMESTICMIG2012  \
1              54660     ...               7.242091         -2.915927   
2             183193     ...              14.832960         17.647293   
3              27341     ...              -4.728132         -2.500690   
69              3169     ...              -5.604858        -12.221204   
70              5546     ...             -11.496318          0.000000   
71            293405     ...              -3.788123         -7.563712   
99             71766     ...               1.775891          1.197250   
100           131809     ...               3.353411        -15.621933   
101           134626     ...             -12.402905          4.908203   
115            19014     ...              -6.544054          2.799271   
116            21831     ...              -8.785244         -9.410783   
117            41538     ...               2.101170          3.230547   
191          1513754     ...              -0.363660          2.488670   
192             1159     ...             -47.577093         11.622709   
193            37873     ...              -3.766029         -8.440854   
250           443680     ...               5.982917          7.689510   
251            15926     ...    

In [44]:
census_df[census_df["SUMLEV"] == 40]["CENSUS2010POP"].head(10)

0       4779736
68       710231
98      6392017
114     2915918
190    37253956
249     5029196
314     3574097
323      897934
327      601723
329    18801310
Name: CENSUS2010POP, dtype: int64

In [45]:
print(census_df.iloc[0])
print(census_df.iloc[68])

print(census_df.iloc[98])

SUMLEV                           40
REGION                            3
DIVISION                          6
STATE                             1
COUNTY                            0
STNAME                      Alabama
CTYNAME                     Alabama
CENSUS2010POP               4779736
ESTIMATESBASE2010           4780127
POPESTIMATE2010             4785161
POPESTIMATE2011             4801108
POPESTIMATE2012             4816089
POPESTIMATE2013             4830533
POPESTIMATE2014             4846411
POPESTIMATE2015             4858979
NPOPCHG_2010                   5034
NPOPCHG_2011                  15947
NPOPCHG_2012                  14981
NPOPCHG_2013                  14444
NPOPCHG_2014                  15878
NPOPCHG_2015                  12568
BIRTHS2010                    14226
BIRTHS2011                    59689
BIRTHS2012                    59062
BIRTHS2013                    57938
BIRTHS2014                    58334
BIRTHS2015                    58305
DEATHS2010                  

First, I have to return the first three states ordered by CENSUS2010POP


In [46]:
census_df[census_df["SUMLEV"] == 40].sort_values(by="CENSUS2010POP", ascending=False)["CENSUS2010POP"].head(10)

190     37253956
2566    25145561
1860    19378102
329     18801310
608     12830632
2282    12702379
2078    11536504
1253     9883640
397      9687653
1923     9535483
Name: CENSUS2010POP, dtype: int64

In [47]:
type(census_df[census_df["SUMLEV"] == 40].sort_values(by="CENSUS2010POP", ascending=False))

pandas.core.frame.DataFrame

In [48]:
test = census_df[census_df["SUMLEV"] == 40].sort_values(by="CENSUS2010POP", ascending=False)
test.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY      STNAME     CTYNAME  \
190       40       4         9      6       0  California  California   
2566      40       3         7     48       0       Texas       Texas   
1860      40       1         2     36       0    New York    New York   
329       40       3         5     12       0     Florida     Florida   
608       40       2         3     17       0    Illinois    Illinois   

      CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
190        37253956           37254503         37334079     ...        
2566       25145561           25146105         25244363     ...        
1860       19378102           19378087         19402920     ...        
329        18801310           18804623         18849890     ...        
608        12830632           12831549         12841249     ...        

      RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
190          -1.148464         -1.163788         -1.339869         -0.862856   
2566          4.527491          5.484904          4.446476          6.072953   
1860         -4.354557         -5.945385         -5.561401         -8.130297   
329           5.556307          5.138184          4.847420          6.958576   
608          -5.243330         -5.805172         -5.308960         -7.509147   

      RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
190          -1.981572     2.761704     2.647127     2.728645     3.743342   
2566          6.248252     7.616364     8.660719     7.739851     9.780487   
1860         -7.990563     1.439239    -0.402860     0.407960    -1.334639   
329          10.080932    11.364173    10.676030    10.565882    13.478570   
608          -8.174675    -3.020410    -3.519051    -2.830379    -4.625216   

      RNETMIG2015  
190      2.656065  
2566     9.978697  
1860    -1.205321  
329     16.528676  
608     -5.247030  

[5 rows x 100 columns]

In [49]:
test = test[["STNAME","CENSUS2010POP"]]
test.head()

STNAME  CENSUS2010POP
190   California       37253956
2566       Texas       25145561
1860    New York       19378102
329      Florida       18801310
608     Illinois       12830632

In [50]:
census_df[census_df["SUMLEV"] == 40].sort_values(by="CENSUS2010POP", ascending=False)[["STNAME","CENSUS2010POP"]].head(3)


STNAME  CENSUS2010POP
190   California       37253956
2566       Texas       25145561
1860    New York       19378102

Now use these three states to get theirs first three counties with higher census population

In [51]:
# Get the three states with most population
test = census_df[census_df["SUMLEV"] == 40].sort_values(by="CENSUS2010POP", ascending=False)[["STNAME","CENSUS2010POP"]].head(3)
#For each one of these three states let's create a list with the three more desnse counties
# normalized_census =census_df[census_df["SUMLEV"] != 40]
three_countries = census_df[census_df["SUMLEV"] != 40].groupby('STNAME').head(3)




In [52]:
test.head()


STNAME  CENSUS2010POP
190   California       37253956
2566       Texas       25145561
1860    New York       19378102

In [53]:
three_countries.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME  \
1       50       3         6      1       1  Alabama   
2       50       3         6      1       3  Alabama   
3       50       3         6      1       5  Alabama   
69      50       4         9      2      13   Alaska   
70      50       4         9      2      16   Alaska   

                       CTYNAME  CENSUS2010POP  ESTIMATESBASE2010  \
1               Autauga County          54571              54571   
2               Baldwin County         182265             182265   
3               Barbour County          27457              27457   
69      Aleutians East Borough           3141               3141   
70  Aleutians West Census Area           5561               5561   

    POPESTIMATE2010     ...       RDOMESTICMIG2011  RDOMESTICMIG2012  \
1             54660     ...               7.242091         -2.915927   
2            183193     ...              14.832960         17.647293   
3             27341     ...              -4.728132         -2.500690   
69             3169     ...              -5.604858        -12.221204   
70             5546     ...             -11.496318          0.000000   

    RDOMESTICMIG2013  RDOMESTICMIG2014  RDOMESTICMIG2015  RNETMIG2011  \
1          -3.012349          2.265971         -2.530799     7.606016   
2          21.845705         19.243287         17.197872    15.844176   
3          -7.056824         -3.904217        -10.543299    -4.874741   
69        -17.257039        -27.748454        -16.221087    19.305620   
70        -13.711875        -11.179039        -26.717891     1.257410   

    RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  
1     -2.626146    -2.722002     2.592270    -2.187333  
2     18.559627    22.727626    20.317142    18.293499  
3     -2.758113    -7.167664    -3.978583   -10.543299  
69     8.249313     5.752346    -3.920977     7.509763  
70    13.314397     0.175793     4.890830   -10.826858  

[5 rows x 100 columns]

In [54]:
list = [three_countries[three_countries["STNAME"] == row["STNAME"]].head(3) for row in test]

TypeError: string indices must be integers

In [ ]:
row = test[test["STNAME"] == "California"]
print(row)


In [ ]:
print(row["STNAME"])

In [ ]:
print(row.get("STNAME").values[0])

In [ ]:
print(row["STNAME"].values[0])

In [ ]:
three_countries["STNAME"].head()

In [ ]:
three_countries[three_countries["STNAME"] == "Alabama"]

In [ ]:
test["STNAME"]

In [ ]:
for row in test.values:
    print(row)

In [ ]:
list = [three_countries[three_countries["STNAME"] == row["STNAME"]].head(3)

Only looking at the three most populous counties for each state, what are the three most populous states (in order of highest population to lowest population)? Use CENSUS2010POP.

This function should return a list of string values.

In [55]:
test

STNAME  CENSUS2010POP
190   California       37253956
2566       Texas       25145561
1860    New York       19378102

In [61]:
type(test.iloc[0])

pandas.core.series.Series

In [65]:
for row in test.values:
    print(row[0]) 
    

California
Texas
New York


In [71]:
list = [three_countries[three_countries["STNAME"] == row[0]].head(3) for row in test]


In [81]:
list = three_countries[three_countries.sort_values(by="CENSUS2010POP", ascending=False)["STNAME"] == "California"].head(3)
type(list)

pandas.core.frame.DataFrame

In [82]:
list.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY      STNAME             CTYNAME  \
209      50       4         9      6      37  California  Los Angeles County   
227      50       4         9      6      73  California    San Diego County   
220      50       4         9      6      59  California       Orange County   

     CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
209        9818605            9818700          9826009     ...        
227        3095313            3095308          3104386     ...        
220        3010232            3010266          3017866     ...        

     RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
209         -4.999439         -4.071266         -4.332318         -5.394299   
227          0.354751          0.235513         -0.246567          0.657077   
220          1.992772          0.618346         -0.491825         -2.374727   

     RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
209         -6.027694    -0.007808     0.377510     0.578049     0.215038   
227         -2.855045     4.382515     5.773072     4.745940     5.921091   
220         -3.182714     6.033306     4.395138     3.681764     2.407607   

     RNETMIG2015  
209    -0.437481  
227     2.770965  
220     1.587398  

[3 rows x 100 columns]

In [78]:
three_countries = census_df[census_df["SUMLEV"] != 40].sort_values(by="CENSUS2010POP", ascending=False).groupby('STNAME').head(3)

In [83]:
three_countries

SUMLEV  REGION  DIVISION  STATE  COUNTY         STNAME  \
209       50       4         9      6      37     California   
624       50       2         3     17      31       Illinois   
2667      50       3         7     48     201          Texas   
106       50       4         8      4      13        Arizona   
227       50       4         9      6      73     California   
220       50       4         9      6      59     California   
1884      50       1         2     36      47       New York   
372       50       3         5     12      86        Florida   
2623      50       3         7     48     113          Texas   
1901      50       1         2     36      81       New York   
1777      50       4         8     32       3         Nevada   
3017      50       4         9     53      33     Washington   
1335      50       2         3     26     163       Michigan   
2786      50       3         7     48     439          Texas   
335       50       3         5     12      11        Florida   
1891      50       1         2     36      61       New York   
2333      50       1         2     42     101   Pennsylvania   
1247      50       1         1     25      17  Massachusetts   
379       50       3         5     12      99        Florida   
2096      50       2         3     39      35           Ohio   
2284      50       1         2     42       3   Pennsylvania   
1316      50       2         3     26     125       Michigan   
2103      50       2         3     39      49           Ohio   
1364      50       2         4     27      53      Minnesota   
2895      50       3         5     51      59       Virginia   
2839      50       4         8     49      35           Utah   
1604      50       2         4     29     189       Missouri   
109       50       4         8      4      19        Arizona   
1228      50       3         5     24      31       Maryland   
559       50       4         9     15       3         Hawaii   
...      ...     ...       ...    ...     ...            ...   
886       50       2         4     19     163           Iowa   
324       50       3         5     10       1       Delaware   
1442      50       3         6     28      33    Mississippi   
1069      50       3         6     21     117       Kentucky   
1758      50       2         4     31     153       Nebraska   
2855      50       1         1     50       7        Vermont   
562       50       4         9     15       9         Hawaii   
1206      50       1         1     23      19          Maine   
2033      50       2         4     38      17   North Dakota   
1680      50       4         8     30     111        Montana   
1800      50       1         1     33      13  New Hampshire   
1853      50       4         8     35      49     New Mexico   
591       50       4         8     16      55          Idaho   
2355      50       1         1     44       9   Rhode Island   
1656      50       4         8     30      63        Montana   
3042      50       3         5     54       3  West Virginia   
2455      50       2         4     46     103   South Dakota   
76        50       4         9      2      90         Alaska   
3046      50       3         5     54      11  West Virginia   
3180      50       4         8     56      21        Wyoming   
1639      50       4         8     30      29        Montana   
85        50       4         9      2     170         Alaska   
2032      50       2         4     38      15   North Dakota   
3182      50       4         8     56      25        Wyoming   
2042      50       2         4     38      35   North Dakota   
2862      50       1         1     50      21        Vermont   
2863      50       1         1     50      23        Vermont   
1792      50       4         8     32     510         Nevada   
3172      50       4         8     56       5        Wyoming   
2444      50       2         4     46      83   South Dakota   

                           CTYNAME  C

In [84]:
list = three_countries[three_countries["STNAME"] == "California"].head(3)
type(list)

pandas.core.frame.DataFrame

In [86]:
list.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY      STNAME             CTYNAME  \
209      50       4         9      6      37  California  Los Angeles County   
227      50       4         9      6      73  California    San Diego County   
220      50       4         9      6      59  California       Orange County   

     CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
209        9818605            9818700          9826009     ...        
227        3095313            3095308          3104386     ...        
220        3010232            3010266          3017866     ...        

     RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
209         -4.999439         -4.071266         -4.332318         -5.394299   
227          0.354751          0.235513         -0.246567          0.657077   
220          1.992772          0.618346         -0.491825         -2.374727   

     RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
209         -6.027694    -0.007808     0.377510     0.578049     0.215038   
227         -2.855045     4.382515     5.773072     4.745940     5.921091   
220         -3.182714     6.033306     4.395138     3.681764     2.407607   

     RNETMIG2015  
209    -0.437481  
227     2.770965  
220     1.587398  

[3 rows x 100 columns]

In [94]:
list = three_countries[three_countries["STNAME"] == "California"].head(3)["CTYNAME"].values
print(list)


['Los Angeles County' 'San Diego County' 'Orange County']


In [102]:
list = [three_countries[three_countries["STNAME"] == row[0]].head(3)["CTYNAME"].values for row in test.values]
print(list)

[array(['Los Angeles County', 'San Diego County', 'Orange County'], dtype=object), array(['Harris County', 'Dallas County', 'Tarrant County'], dtype=object), array(['Kings County', 'Queens County', 'New York County'], dtype=object)]


In [101]:
three_countries[three_countries["STNAME"] == "California"].head(3)["CTYNAME"].values

array(['Los Angeles County', 'San Diego County', 'Orange County'], dtype=object)

In [110]:
def answer_six():
    states = census_df[census_df["SUMLEV"] == 40].sort_values(by="CENSUS2010POP", \
            ascending=False)[["STNAME","CENSUS2010POP"]].head(3)
    counties = census_df[census_df["SUMLEV"] != 40].sort_values(by="CENSUS2010POP", \
            ascending=False).groupby('STNAME').head(3)
    top_counties = [counties[counties["STNAME"] == state[0]].head(3)["CTYNAME"].values for state in states.values]
    result = []
    for item in top_counties:
        result.extend(item)
    return result

In [112]:
answer_six()


['Los Angeles County',
 'San Diego County',
 'Orange County',
 'Harris County',
 'Dallas County',
 'Tarrant County',
 'Kings County',
 'Queens County',
 'New York County']

In [116]:
census_df.values.T

array([[40, 50, 50, ..., 50, 50, 50],
       [3, 3, 3, ..., 4, 4, 4],
       [6, 6, 6, ..., 8, 8, 8],
       ..., 
       [1.3832821479, -2.722001942, 22.727626439, ..., -7.521839517999999,
        -2.250384934, 12.37258348],
       [1.7247180515, 2.5922703212, 20.317141924, ..., -14.74060823,
        -18.02016827, 1.533635413],
       [0.7125937237000001, -2.187333351, 18.293499115, ..., -12.60635111,
        1.4419610671, 6.935293709700001]], dtype=object)

In [118]:
census_df.columns

Index(['SUMLEV', 'REGION', 'DIVISION', 'STATE', 'COUNTY', 'STNAME', 'CTYNAME',
       'CENSUS2010POP', 'ESTIMATESBASE2010', 'POPESTIMATE2010',
       'POPESTIMATE2011', 'POPESTIMATE2012', 'POPESTIMATE2013',
       'POPESTIMATE2014', 'POPESTIMATE2015', 'NPOPCHG_2010', 'NPOPCHG_2011',
       'NPOPCHG_2012', 'NPOPCHG_2013', 'NPOPCHG_2014', 'NPOPCHG_2015',
       'BIRTHS2010', 'BIRTHS2011', 'BIRTHS2012', 'BIRTHS2013', 'BIRTHS2014',
       'BIRTHS2015', 'DEATHS2010', 'DEATHS2011', 'DEATHS2012', 'DEATHS2013',
       'DEATHS2014', 'DEATHS2015', 'NATURALINC2010', 'NATURALINC2011',
       'NATURALINC2012', 'NATURALINC2013', 'NATURALINC2014', 'NATURALINC2015',
       'INTERNATIONALMIG2010', 'INTERNATIONALMIG2011', 'INTERNATIONALMIG2012',
       'INTERNATIONALMIG2013', 'INTERNATIONALMIG2014', 'INTERNATIONALMIG2015',
       'DOMESTICMIG2010', 'DOMESTICMIG2011', 'DOMESTICMIG2012',
       'DOMESTICMIG2013', 'DOMESTICMIG2014', 'DOMESTICMIG2015', 'NETMIG2010',
       'NETMIG2011', 'NETMIG2012', 'NETMI

In [121]:
df1 = pd.DataFrame(census_df.values.T, index=census_df.columns)

In [123]:
df1

0               1               2     \
SUMLEV                         40              50              50   
REGION                          3               3               3   
DIVISION                        6               6               6   
STATE                           1               1               1   
COUNTY                          0               1               3   
STNAME                    Alabama         Alabama         Alabama   
CTYNAME                   Alabama  Autauga County  Baldwin County   
CENSUS2010POP             4779736           54571          182265   
ESTIMATESBASE2010         4780127           54571          182265   
POPESTIMATE2010           4785161           54660          183193   
POPESTIMATE2011           4801108           55253          186659   
POPESTIMATE2012           4816089           55175          190396   
POPESTIMATE2013           4830533           55038          195126   
POPESTIMATE2014           4846411           55290          199713   
POPESTIMATE2015           4858979           55347          203709   
NPOPCHG_2010                 5034              89             928   
NPOPCHG_2011                15947             593            3466   
NPOPCHG_2012                14981             -78            3737   
NPOPCHG_2013                14444            -137            4730   
NPOPCHG_2014                15878             252            4587   
NPOPCHG_2015                12568              57            3996   
BIRTHS2010                  14226             151             517   
BIRTHS2011                  59689             636            2187   
BIRTHS2012                  59062             615            2092   
BIRTHS2013                  57938             574            2160   
BIRTHS2014                  58334             623            2186   
BIRTHS2015                  58305             600            2240   
DEATHS2010                  11089             152             532   
DEATHS2011                  48811             507            1825   
DEATHS2012                  48357             558            1879   
...                           ...             ...             ...   
RBIRTH2011                 12.453         11.5728         11.8264   
RBIRTH2012                12.2826         11.1385         11.0965   
RBIRTH2013                12.0121         10.4162         11.2056   
RBIRTH2014                12.0563         11.2936         11.0729   
RBIRTH2015                 12.015         10.8463          11.105   
RDEATH2011                10.1835         9.22548         9.86881   
RDEATH2012                10.0564         10.1061         9.96672   
RDEATH2013                10.5411         10.5795         9.86714   
RDEATH2014                 10.381         9.13639         10.3536   
RDEATH2015                10.3716         8.44202         9.87551   
RNATURALINC2011            2.2695         2.34731         1.95754   
RNATURALINC2012           2.22622         1.03235         1.12981   
RNATURALINC2013           1.47098        -0.16332         1.33845   
RNATURALINC2014           1.67532          2.1572        0.719281   
RNATURALINC2015           1.64342         2.40426         1.22948   
RINTERNATIONALMIG2011     1.02772        0.363924         1.01122   
RINTERNATIONALMIG2012     1.01984        0.289782        0.912334   
RINTERNATIONALMIG2013     1.00222        0.290347        0.881921   
RINTERNATIONALMIG2014     1.14272          0.3263         1.07386   
RINTERNATIONALMIG2015     1.17996        0.343466         1.09563   
RDOMESTICMIG2011       0.00229495         7.24209          14.833   
RDOMESTICMIG2012        -0.193196        -2.91593         17.6473   
RDOMESTICMIG2013         0.381066        -3.01235         21.8457   
RDOMESTICMIG2014         0.582002         2.26597         19.2433   
RDOMESTICMIG2015        -0.467369         -2.5308         17.1979   
RNETMIG2011               1.03001         7.60602         15.8442   
RNETMIG2012              0.

In [132]:
min_pop = census_df.loc[:, "POPESTIMATE2010":"POPESTIMATE2015"]

In [133]:
min_pop.head()

POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
0          4785161          4801108          4816089          4830533   
1            54660            55253            55175            55038   
2           183193           186659           190396           195126   
3            27341            27226            27159            26973   
4            22861            22733            22642            22512   

   POPESTIMATE2014  POPESTIMATE2015  
0          4846411          4858979  
1            55290            55347  
2           199713           203709  
3            26815            26489  
4            22549            22583

In [136]:
min_data = min_pop.min(axis=1)
max_data = min_pop.max(axis=1)
rest_data = max_data - min_data

In [137]:
min_data.head()

0    4785161
1      54660
2     183193
3      26489
4      22512
dtype: int64

In [139]:
max_data.head()

0    4858979
1      55347
2     203709
3      27341
4      22861
dtype: int64

In [141]:
rest_data.head()

0    73818
1      687
2    20516
3      852
4      349
dtype: int64

In [144]:
max_rest = rest_data.head().argmax()
print(max_rest)

0


In [146]:
max_rest = rest_data.head().max()
print(max_rest)

73818


In [151]:
census = census_df.loc[rest_data.head().argmax(),"CTYNAME"]

In [154]:
print(census)

Alabama


In [158]:
remove_states = min_pop = census_df[census_df["SUMLEV"] != 40]

In [159]:
min_pop = remove_states.loc[:, "POPESTIMATE2010":"POPESTIMATE2015"]

In [160]:
min_data = min_pop.min(axis=1)
max_data = min_pop.max(axis=1)
rest_data = max_data - min_data

In [162]:
census = census_df.loc[rest_data.head().argmax(),"CTYNAME"]

In [164]:
print(census)

Baldwin County


In [166]:
census_df[census_df["CTYNAME"] == "Baldwin County"]

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
2        50       3         6      1       3  Alabama  Baldwin County   
402      50       3         5     13       9  Georgia  Baldwin County   

     CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
2           182265             182265           183193     ...        
402          45720              45835            45702     ...        

     RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
2            14.83296         17.647293         21.845705         19.243287   
402         -17.57748         26.847544         -9.633245         -8.721183   

     RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
2           17.197872    15.844176    18.559627    22.727626    20.317142   
402         -8.786056   -16.828565    27.590275    -8.769277    -7.742497   

     RNETMIG2015  
2      18.293499  
402    -7.800090  

[2 rows x 100 columns]

In [241]:
def answer_seven():
    population = census_df[census_df["SUMLEV"] != 40].loc[:, "POPESTIMATE2010":"POPESTIMATE2015"]
    rest_data = population.max(axis=1) -  min_pop.min(axis=1)
    return census_df.loc[rest_data.argmax(),"CTYNAME"]

In [242]:
answer_seven()

'Harris County'

In [179]:
census_df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
0      40       3         6      1       0  Alabama         Alabama   
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
0        4779736            4780127          4785161     ...        
1          54571              54571            54660     ...        
2         182265             182265           183193     ...        
3          27457              27457            27341     ...        
4          22915              22919            22861     ...        

   RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
0          0.002295         -0.193196          0.381066          0.582002   
1          7.242091         -2.915927         -3.012349          2.265971   
2         14.832960         17.647293         21.845705         19.243287   
3         -4.728132         -2.500690         -7.056824         -3.904217   
4         -5.527043         -5.068871         -6.201001         -0.177537   

   RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
0         -0.467369     1.030015     0.826644     1.383282     1.724718   
1         -2.530799     7.606016    -2.626146    -2.722002     2.592270   
2         17.197872    15.844176    18.559627    22.727626    20.317142   
3        -10.543299    -4.874741    -2.758113    -7.167664    -3.978583   
4          0.177258    -5.088389    -4.363636    -5.403729     0.754533   

   RNETMIG2015  
0     0.712594  
1    -2.187333  
2    18.293499  
3   -10.543299  
4     1.107861  

[5 rows x 100 columns]

### Question 8
In this datafile, the United States is broken up into four regions using the "REGION" column. 

Create a query that finds the counties that belong to regions 1 or 2, whose name starts with 'Washington', and whose POPESTIMATE2015 was greater than their POPESTIMATE 2014.

*This function should return a 5x2 DataFrame with the columns = ['STNAME', 'CTYNAME'] and the same index ID as the census_df (sorted ascending by index).*

In [206]:
first_filter = census_df[census_df["REGION"].isin([1,2])]
first_filter.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY       STNAME            CTYNAME  \
314      40       1         1      9       0  Connecticut        Connecticut   
315      50       1         1      9       1  Connecticut   Fairfield County   
316      50       1         1      9       3  Connecticut    Hartford County   
317      50       1         1      9       5  Connecticut  Litchfield County   
318      50       1         1      9       7  Connecticut   Middlesex County   

     CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
314        3574097            3574118          3579717     ...        
315         916829             916850           919744     ...        
316         894014             894029           895303     ...        
317         189927             189927           189773     ...        
318         165676             165676           165628     ...        

     RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
314         -3.724400         -5.311208         -4.730271         -7.567093   
315         -0.228261         -3.030708         -2.070291         -6.633789   
316         -5.037297         -6.324143         -5.237664         -8.636074   
317         -4.103317         -7.183847         -2.812353         -7.951141   
318          1.410378         -6.716791         -2.617825         -4.987002   

     RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
314         -7.687268     0.775510    -1.066084    -0.278693    -2.376831   
315         -7.385939     6.119788     2.687406     4.096896     0.441829   
316         -8.135161    -0.025666    -1.662136    -0.314193    -2.840433   
317         -5.931948    -3.290047    -6.450584    -1.983618    -6.939080   
318         -6.526367     3.598271    -4.522077    -0.308335    -2.272328   

     RNETMIG2015  
314    -2.463243  
315    -0.338989  
316    -2.327200  
317    -4.906204  
318    -3.807554  

[5 rows x 100 columns]

In [209]:
second_filter = first_filter[first_filter["CTYNAME"].str.startswith("Washington")]
second_filter

SUMLEV  REGION  DIVISION  STATE  COUNTY        STNAME  \
703       50       2         3     17     189      Illinois   
799       50       2         3     18     175       Indiana   
896       50       2         4     19     183          Iowa   
1005      50       2         4     20     201        Kansas   
1211      50       1         1     23      29         Maine   
1419      50       2         4     27     163     Minnesota   
1618      50       2         4     29     221      Missouri   
1770      50       2         4     31     177      Nebraska   
1918      50       1         2     36     115      New York   
2162      50       2         3     39     167          Ohio   
2345      50       1         2     42     125  Pennsylvania   
2355      50       1         1     44       9  Rhode Island   
2863      50       1         1     50      23       Vermont   
3163      50       2         3     55     131     Wisconsin   

                CTYNAME  CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  \
703   Washington County          14716              14716            14701   
799   Washington County          28262              28262            28284   
896   Washington County          21704              21704            21697   
1005  Washington County           5799               5799             5782   
1211  Washington County          32856              32856            32824   
1419  Washington County         238136             238134           238995   
1618  Washington County          25195              25196            25177   
1770  Washington County          20234              20234            20275   
1918  Washington County          63216              63216            63292   
2162  Washington County          61778              61778            61708   
2345  Washington County         207820             207820           207877   
2355  Washington County         126979             127094           127085   
2863  Washington County          59534              59526            59570   
3163  Washington County         131887             131885           131967   

         ...       RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  \
703      ...              -8.057633          2.192006        -12.192182   
799      ...              -3.683894         -9.517699         -5.310179   
896      ...               5.743693          2.468684         -0.364282   
1005     ...              10.504563        -18.164519        -18.821460   
1211     ...               0.213626         -3.280699         -4.262285   
1419     ...               2.555118          3.260957          2.285952   
1618     ...              -5.130755         -2.389867          0.079577   
1770     ...              -1.874784         -1.331459         -4.492607   
1918     ...              -3.719090         -1.776030         -4.165839   
2162     ...               0.129798          0.780050         -0.455611   
2345     ...               3.683135          3.260434          1.219337   
2355     ...              -4.669304         -2.659574          1.551904   
2863     ...              -1.578161         -5.617505         -3.407759   
3163     ...              -0.794876          0.785279         -2.215465   

      RDOMESTICMIG2014  RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  \
703          -2.775850        -10.605638    -8.125918     2.123506   
799           0.717901         -5.454774    -3.258829    -9.161231   
896           0.272171          1.849596     5.881542     2.560117   
1005         -5.501819         -6.948156    12.054417   -16.780555   
1211         -9.111049         -4.284274     0.366217    -2.790127   
1419          3.072543          0.678755     4.178076     4.907905   
1618         -1.592420        -11.511311    -5.011435    -2.389867   
1770         -0.889284         -3.998716    -1.825448    -1.282146   
1918         -4.954213         -2.342897    -3.418398    -1.522311   
2162         -2.137467          0.016355     0.227147     1.023

In [202]:
third_filter = second_filter[second_filter["POPESTIMATE2015"] > second_filter["POPESTIMATE2014"]]
third_filter.loc[:,['STNAME', 'CTYNAME']]


STNAME            CTYNAME
896           Iowa  Washington County
1419     Minnesota  Washington County
2345  Pennsylvania  Washington County
2355  Rhode Island  Washington County
3163     Wisconsin  Washington County

In [210]:
def answer_eight():
    first_filter = census_df[census_df["REGION"].isin([1,2])]
    second_filter = first_filter[first_filter["CTYNAME"].str.startswith("Washington")]
    third_filter = second_filter[second_filter["POPESTIMATE2015"] > second_filter["POPESTIMATE2014"]]
    return third_filter.loc[:,['STNAME', 'CTYNAME']]

In [231]:
answer_eight()


STNAME            CTYNAME
896           Iowa  Washington County
1419     Minnesota  Washington County
2345  Pennsylvania  Washington County
2355  Rhode Island  Washington County
3163     Wisconsin  Washington County

In [233]:
    population = census_df[census_df["SUMLEV"] != 40].loc[:, "POPESTIMATE2010":"POPESTIMATE2015"]

In [235]:
population

POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
1               54660            55253            55175            55038   
2              183193           186659           190396           195126   
3               27341            27226            27159            26973   
4               22861            22733            22642            22512   
5               57373            57711            57776            57734   
6               10887            10629            10606            10628   
7               20944            20673            20408            20261   
8              118437           117768           117286           116575   
9               34098            33993            34075            34153   
10              25976            26080            26023            26084   
11              43665            43739            43697            43795   
12              13841            13593            13543            13378   
13              25767            25570            25144            25116   
14              13880            13670            13456            13467   
15              14973            14971            14921            15028   
16              50177            50448            51173            50755   
17              54514            54443            54472            54471   
18              13208            13121            12996            12875   
19              11758            11348            11195            11059   
20              37796            38060            37818            37830   
21              13853            13896            13951            13932   
22              80473            80469            80374            80756   
23              50358            50109            50324            49833   
24              43803            43178            42777            42021   
25              71142            71387            70942            70869   
26              79465            80012            80432            80883   
27              38309            38213            38034            37857   
28             104442           104236           104235           103852   
29              17231            17062            16960            16857   
30              31734            31729            31648            31507   
...               ...              ...              ...              ...   
3162            15930            15784            15831            15657   
3163           131967           132225           132649           132758   
3164           390076           390808           392710           394025   
3165            52422            52342            52035            52213   
3166            24506            24581            24484            24332   
3167           167059           167630           168717           169486   
3168            74807            74647            74384            73996   
3170            36428            36908            37396            37647   
3171            11672            11745            11785            12002   
3172            46244            46600            47881            48121   
3173            15837            15817            15678            15787   
3174            13826            13728            14025            14343   
3175             7114             7129             7148             7160   
3176            40222            40591            41129            41024   
3177            13408            13597            13666            13565   
3178             4813             4818             4846             4846   
3179             8581             8636             8610             8619   
3180            92271            92663            94894            96006   
3181            18091            18022            17943            18328   
3182            75472            76420            78699            81156   
3183             2492             2485             247

In [237]:
    population = census_df[census_df["SUMLEV"] != 40].loc[:, "POPESTIMATE2010":"POPESTIMATE2015"]
    rest_data = population.max(axis=1) -  population.min(axis=1)
    rest_data


1         687
2       20516
3         852
4         349
5         403
6         281
7         790
8        2817
9         160
10        225
11        278
12        671
13       1092
14        424
15        151
16       1034
17        160
18        546
19       1034
20        264
21        110
22       1631
23        857
24       2672
25        518
26       2003
27        525
28       1385
29        472
30        227
        ...  
3162      378
3163     1707
3164     6412
3165      477
3166      548
3167     2580
3168     1372
3170     1528
3171      350
3172     2976
3173      297
3174      615
3175      330
3176      907
3177      283
3178      105
3179       84
3180     4850
3181      779
3182     6706
3183       73
3184      978
3185      134
3186      874
3187      519
3188     1569
3189     1828
3190      280
3191      229
3192      169
dtype: int64

In [239]:
population.max(axis=1)


1        55347
2       203709
3        27341
4        22861
5        57776
6        10887
7        20944
8       118437
9        34153
10       26084
11       43943
12       13841
13       25767
14       13880
15       15072
16       51211
17       54514
18       13208
19       11758
20       38060
21       13963
22       82005
23       50358
24       43803
25       71387
26       81468
27       38309
28      104442
29       17231
30       31734
         ...  
3162     15930
3163    133674
3164    396488
3165     52422
3166     24581
3167    169639
3168     74807
3170     37956
3171     12022
3172     49220
3173     15856
3174     14343
3175      7444
3176     41129
3177     13666
3178      4846
3179      8636
3180     97121
3181     18722
3182     82178
3183      2548
3184     29237
3185      8812
3186     30020
3187     10418
3188     45162
3189     23125
3190     21102
3191      8545
3192      7234
dtype: int64

In [244]:
population.min(axis=1)

1        54660
2       183193
3        26489
4        22512
5        57373
6        10606
7        20154
8       115620
9        33993
10       25859
11       43665
12       13170
13       24675
14       13456
15       14921
16       50177
17       54354
18       12662
19       10724
20       37796
21       13853
22       80374
23       49501
24       41131
25       70869
26       79465
27       37784
28      103057
29       16759
30       31507
         ...  
3162     15552
3163    131967
3164    390076
3165     51945
3166     24033
3167    167059
3168     73435
3170     36428
3171     11672
3172     46244
3173     15559
3174     13728
3175      7114
3176     40222
3177     13383
3178      4741
3179      8552
3180     92271
3181     17943
3182     75472
3183      2475
3184     28259
3185      8678
3186     29146
3187      9899
3188     43593
3189     21297
3190     20822
3191      8316
3192      7065
dtype: int64

I have missunderstood question #6

Question 6

Only looking at the three most populous counties for each state, what are the three most populous states (in order of highest population to lowest population)? Use CENSUS2010POP.

This function should return a list of string values.


In [246]:
def answer_six():
    states = census_df[census_df["SUMLEV"] == 40].sort_values(by="CENSUS2010POP", \
            ascending=False)[["STNAME","CENSUS2010POP"]].head(3)
    counties = census_df[census_df["SUMLEV"] != 40].sort_values(by="CENSUS2010POP", \
            ascending=False).groupby('STNAME').head(3)
    top_counties = [counties[counties["STNAME"] == state[0]].head(3)["CTYNAME"].values for state in states.values]
    result = []
    for item in top_counties:
        result.extend(item)
    return result

In [250]:
# These are the three countries with higes population
# I have removed the sumlev that belong to the state it self
counties = census_df[census_df["SUMLEV"] != 40].sort_values(by="CENSUS2010POP", \
            ascending=False).groupby('STNAME').head(3)
counties.sort_values(by="STNAME").head(10)

SUMLEV  REGION  DIVISION  STATE  COUNTY    STNAME  \
45       50       3         6      1      89   Alabama   
37       50       3         6      1      73   Alabama   
49       50       3         6      1      97   Alabama   
71       50       4         9      2      20    Alaska   
85       50       4         9      2     170    Alaska   
76       50       4         9      2      90    Alaska   
106      50       4         8      4      13   Arizona   
110      50       4         8      4      21   Arizona   
109      50       4         8      4      19   Arizona   
118      50       3         7      5       7  Arkansas   

                          CTYNAME  CENSUS2010POP  ESTIMATESBASE2010  \
45                 Madison County         334811             334811   
37               Jefferson County         658466             658350   
49                  Mobile County         412992             413143   
71         Anchorage Municipality         291826             291826   
85      Matanuska-Susitna Borough          88995              88995   
76   Fairbanks North Star Borough          97581              97581   
106               Maricopa County        3817117            3817357   
110                  Pinal County         375770             375770   
109                   Pima County         980263             980263   
118                 Benton County         221339             221344   

     POPESTIMATE2010     ...       RDOMESTICMIG2011  RDOMESTICMIG2012  \
45            336138     ...               4.170416          3.438521   
37            658116     ...              -4.138628         -4.492646   
49            413332     ...              -5.242052         -3.722733   
71            293405     ...              -3.788123         -7.563712   
85             89761     ...              13.050517         11.474315   
76             98289     ...             -17.339069          5.430293   
106          3825597     ...               2.444413          9.232973   
110           379384     ...             -13.828377          4.308298   
109           981870     ...               0.874159         -0.248360   
118           223010     ...              11.164115         11.786353   

     RDOMESTICMIG2013  RDOMESTICMIG2014  RDOMESTICMIG2015  RNETMIG2011  \
45           6.367708          4.064107          2.480087     5.739131   
37          -1.664873         -0.970225         -3.337536    -2.985041   
49          -3.181655         -2.311087         -3.873201    -3.751237   
71          -3.962343        -18.897015        -19.781254    -0.969922   
85          12.906763         15.008086         18.697964    13.391923   
76          -9.147023        -33.436804        -13.859157   -15.791483   
106          7.879873          8.318210          9.123309     4.726436   
110          0.966512         17.437436         18.998761   -11.594501   
109         -1.381638          1.036139          0.159859     2.781876   
118          9.333699         11.268649         16.254517    13.816535   

     RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  
45      4.880766     7.818211     5.734524     4.180881  
37     -3.536344    -0.647113     0.221333    -2.126089  
49     -2.281051    -1.753896    -0.759909    -2.288272  
71     -2.652342     0.053320   -14.789980   -15.280810  
85     12.303913    13.528903    15.605525    19.349985  
76     10.618791    -4.926084   -29.823633    -8.984834  
106    11.527593    10.390177    11.203958    11.985998  
110     6.465078     3.391910    20.212795    21.745600  
109     2.000003     0.854104     3.582023     2.777186  
118    14.533319    12.296171    14.648828    19.555772  

[10 rows x 100 columns]

Now I think I have to group by stname and do the sum

In [259]:
counties[counties["STNAME"] == "Alabama"]

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME           CTYNAME  \
37      50       3         6      1      73  Alabama  Jefferson County   
49      50       3         6      1      97  Alabama     Mobile County   
45      50       3         6      1      89  Alabama    Madison County   

    CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
37         658466             658350           658116     ...        
49         412992             413143           413332     ...        
45         334811             334811           336138     ...        

    RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
37         -4.138628         -4.492646         -1.664873         -0.970225   
49         -5.242052         -3.722733         -3.181655         -2.311087   
45          4.170416          3.438521          6.367708          4.064107   

    RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
37         -3.337536    -2.985041    -3.536344    -0.647113     0.221333   
49         -3.873201    -3.751237    -2.281051    -1.753896    -0.759909   
45          2.480087     5.739131     4.880766     7.818211     5.734524   

    RNETMIG2015  
37    -2.126089  
49    -2.288272  
45     4.180881  

[3 rows x 100 columns]

In [258]:
counties.groupby('STNAME').sum()

SUMLEV  REGION  DIVISION  STATE  COUNTY  CENSUS2010POP  \
STNAME                                                                         
Alabama                  150       9        18      3     259        1406269   
Alaska                   150      12        27      6     280         478402   
Arizona                  150      12        24     12      53        5173150   
Arkansas                 150       9        21     15     269         807152   
California               150      12        27     18     169       15924150   
Colorado                 150      12        24     24      77        1794424   
Connecticut              150       3         3     27      13        2673320   
Delaware                 150       9        15     30       9         897934   
District of Columbia      50       3         5     11       1         601723   
Florida                  150       9        15     36     196        5564635   
Georgia                  150       9        15     39     345        2417795   
Hawaii                   150      12        27     45      13        1293120   
Idaho                    150      12        24     48      83         719782   
Illinois                 150       6         9     51     171        6815061   
Indiana                  150       6         9     54     189        1754727   
Iowa                     150       6        12     57     429         807090   
Kansas                   150       6        12     60     441        1220478   
Kentucky                 150       9        18     63     295        1196619   
Louisiana                150       9        21     66     155        1216552   
Maine                    150       3         3     69      55         632728   
Maryland                 150       9        15     72      69        2640226   
Massachusetts            150       3         3     75      53        3044796   
Michigan                 150       6         9     78     387        3863924   
Minnesota                150       6        12     81     213        2059617   
Mississippi              150       9        18     84     129         593642   
Missouri                 150       6        12     87     467        2033597   
Montana                  150      12        24     90     203         348199   
Nebraska                 150       6        12     93     317         961357   
Nevada                   150      12        24     96     544        2427950   
New Hampshire            150       3         3     99      39         842389   
New Jersey               150       3         6    102      39        2498943   
New Mexico               150      12        24    105      63        1015967   
New York                 150       3         6    108     189        6321295   
North Carolina           150       9        15    111     383        2309027   
North Dakota             150       6        12    114      67         297947   
Ohio                     150       6         9    117     145        3245910   
Oklahoma                 150       9        21    120     279        1577791   
Oregon                   150      12        27    123     123        1641036   
Pennsylvania             150       3         6    126     195        3549228   
Rhode Island             150       3         3    132      19         919804   
South Carolina           150       9        15    135     143        1185938   
South Dakota             150       6        12    138     285         315244   
Tennessee                150       9        18    141     287        1986551   
Texas                    150       9        21    144     753        8269632   
Utah                     150      12        24    147      95        1852698   
Vermont                  150       3         3    150      51         277721   
Virginia                 150       9        15    153    1022        1921722   
Washington               150      12        27    159     147        3439809   
West Virginia         

In [261]:
def answer_six():
    counties = census_df[census_df["SUMLEV"] != 40].sort_values(by="CENSUS2010POP", \
            ascending=False).groupby('STNAME').head(3).groupby('STNAME').sum()
    return counties.sort_values(by="CENSUS2010POP", ascending=False)

In [263]:
answer_six()

SUMLEV  REGION  DIVISION  STATE  COUNTY  CENSUS2010POP  \
STNAME                                                                         
California               150      12        27     18     169       15924150   
Texas                    150       9        21    144     753        8269632   
Illinois                 150       6         9     51     171        6815061   
New York                 150       3         6    108     189        6321295   
Florida                  150       9        15     36     196        5564635   
Arizona                  150      12        24     12      53        5173150   
Michigan                 150       6         9     78     387        3863924   
Pennsylvania             150       3         6    126     195        3549228   
Washington               150      12        27    159     147        3439809   
Ohio                     150       6         9    117     145        3245910   
Massachusetts            150       3         3     75      53        3044796   
Connecticut              150       3         3     27      13        2673320   
Maryland                 150       9        15     72      69        2640226   
New Jersey               150       3         6    102      39        2498943   
Nevada                   150      12        24     96     544        2427950   
Georgia                  150       9        15     39     345        2417795   
North Carolina           150       9        15    111     383        2309027   
Minnesota                150       6        12     81     213        2059617   
Missouri                 150       6        12     87     467        2033597   
Tennessee                150       9        18    141     287        1986551   
Virginia                 150       9        15    153    1022        1921722   
Utah                     150      12        24    147      95        1852698   
Wisconsin                150       6         9    165     237        1825699   
Colorado                 150      12        24     24      77        1794424   
Indiana                  150       6         9     54     189        1754727   
Oregon                   150      12        27    123     123        1641036   
Oklahoma                 150       9        21    120     279        1577791   
Alabama                  150       9        18      3     259        1406269   
Hawaii                   150      12        27     45      13        1293120   
Kansas                   150       6        12     60     441        1220478   
Louisiana                150       9        21     66     155        1216552   
Kentucky                 150       9        18     63     295        1196619   
South Carolina           150       9        15    135     143        1185938   
New Mexico               150      12        24    105      63        1015967   
Nebraska                 150       6        12     93     317         961357   
Rhode Island             150       3         3    132      19         919804   
Delaware                 150       9        15     30       9         897934   
New Hampshire            150       3         3     99      39         842389   
Arkansas                 150       9        21     15     269         807152   
Iowa                     150       6        12     57     429         807090   
Idaho                    150      12        24     48      83         719782   
Maine                    150       3         3     69      55         632728   
District of Columbia      50       3         5     11       1         601723   
Mississippi              150       9        18     84     129         593642   
Alaska                   150      12        27      6     280         478402   
West Virginia            150       9        15    162      53         393551   
Montana                  150      12        24     90     203         348199   
South Dakota             150       6        12    138     285         315244   
North Dakota          

In [270]:
def answer_six():
    counties = census_df[census_df["SUMLEV"] != 40].sort_values(by="CENSUS2010POP", \
            ascending=False).groupby('STNAME').head(3).groupby('STNAME').sum()
    return counties.sort_values(by="CENSUS2010POP", ascending=False).head(3).index.values

In [273]:
answer_six()

array(['California', 'Texas', 'Illinois'], dtype=object)

In [274]:
type(answer_six())

numpy.ndarray

In [282]:
def answer_six():
    counties = census_df[census_df["SUMLEV"] != 40].sort_values(by="CENSUS2010POP", \
            ascending=False).groupby('STNAME').head(3).groupby('STNAME').sum()
    return counties.sort_values(by="CENSUS2010POP", ascending=False).head(3)

In [290]:
result = answer_six()
result.index.values.tolist()

['California', 'Texas', 'Illinois']

In [289]:
result['STNAME']

KeyError: 'STNAME'